### PR TITLE
Adds tests for `Q8` and `Q4`

### PR DESCRIPTION
**Pull request by juanjqo**
@dqrobotics/developers 

Hi @mmmarinho and @bvadorno 

This PR adds the tests for Q8 operations. I updated the file `DQ_test.m` to generate new data containing the Q8 tests and saved the results in `DQ_test.mat`. The legacy data was renamed to `DQ_test_october_10_2019.mat`. Furthermore, I updated the file `DQ_test.py` to include a new test: `test_of_Q8`.

However, the `test_ad` in `DQ_test.py` fails with the new data 'DQ_test.mat' with the following assertion:

```shell
AssertionError: 
Arrays are not almost equal to 12 decimals
Error in Ad
Mismatched elements: 1 / 8 (12.5%)
Max absolute difference: 1.8189894e-12
Max relative difference: 5.05591348e-16
```
Because of that, I am using `DQ_test_october_10_2019.mat` (legacy data) for all tests except for
`test_of_Q8`, which uses 'DQ_test.mat' (the new data).

Best regards, 

Juancho

### CODE REVIEW

**Comment by mmmarinho** [comment]
@juanjqo 

Thanks! 

Can you see any obvious pattern on the data that stopped working? 

For instance, one of the numbers is a real number, or a pure quaternion, or so on.

**Comment by juanjqo** [comment]
@mmmarinho not yet. I did some tests in Matlab.  When I use the `legacy data` within an absolute tolerance of `esp` (2.2204e-16) the test fails with an error of `2.2737e-13` (`--> The error was not within absolute tolerance.)`

This table containts the results using two tolerances and both data files.

| File  | Data Status | Tolerance | Error | Test |
| ------------- | ------------- | ------------- | ------------- | ------------- |
| DQ_test_october_10_2019.mat  | Legacy  | eps | 2.2737e-13 | fails |
| DQ_test_october_10_2019.mat  | Legacy  | DQ.threshold | -- | OK |
| DQ_test.mat  | Current  | eps | -- | OK |
| DQ_test.mat  | Current  | DQ.threshold | -- | OK |

Test:
```matlab
classdef DQ_CustomTest < matlab.unittest.TestCase
        properties
        tolerance  = eps;  % eps= 2.2204e-16;
        end
    methods (Test)
            function test_ad(test_case)
                 import matlab.unittest.constraints.IsEqualTo;
                 import matlab.unittest.constraints.AbsoluteTolerance;
                 load('C:\Users\juanjqo\Documents\git\DQ_Robotics\python-tests\DQ_test_october_10_2019.mat');

                 for i=1:NUMBER_OF_RANDOM
                        a = DQ(random_dq_a(:,i));
                        b = DQ(random_dq_b(:,i));
                        expected_solution = result_of_Ad(:,i);
                        actual_solution = vec8(Ad(a,b));

                        test_case.verifyThat(actual_solution,IsEqualTo(expected_solution,...
                        'Within', AbsoluteTolerance(test_case.tolerance))); 
                 end
            end
    end
end
```

I'm going to perform some tests in C++. If there is some pattern, it is not obvious to me.

Best regards, 

Juancho

**Comment by mmmarinho** [comment]
@juanjqo 
Thanks! Can you remind me why we're not testing `Q4` as well?

**Comment by juanjqo** [comment]
Hi @mmmarinho, 

The method Q4() is unavailable for the user in `dqrobotics/matlab`. Therefore, I did not generate the required data (DQ_test.mat) to perform the test in `DQ_test.py`. 

```matlab
>> Q8(DQ(1))

ans =

     0     0     0     0     0     0
     1     0     0     0     0     0
     0     1     0     0     0     0
     0     0     1     0     0     0
     0     0     0     0     0     0
     0     0     0     1     0     0
     0     0     0     0     1     0
     0     0     0     0     0     1
```
But
```matlab
>> Q4(DQ(1))
Unrecognized function or variable 'Q4'.
```

**Comment by mmmarinho** [comment]
@juanjqo If Q4 isn’t available in MATLAB exposing it in CPP trades one language incompatibility for another. Let’s fix that first.

**Comment by juanjqo** [comment]
Hi @mmmarinho I added the tests for Q4! :neckbeard:

Cheers, 

Juancho

File: DQ_test.py [hunk 1]
```diff
@@ -24,8 +24,27 @@
...
+Arrays are not almost equal to 12 decimals
+Error in Ad
+Mismatched elements: 1 / 8 (12.5%)
+Max absolute difference: 1.8189894e-12
+Max relative difference: 5.05591348e-16
+
+This is the reason to use 'DQ_test_october_10_2019.mat' for all tests except 
+test_of_Q4 and test_of_Q8, which use 'DQ_test.mat.
+"""
+mat = scipy.io.loadmat('DQ_test_october_10_2019.mat')
```
Comment (mmmarinho) [inline DQ_test.py]: @juanjqo

1.  I suppose that eventually we'll need to parse the name of these `.mat` files, so we can make them a little more computer friendly. For instance, rename to `DQ_test_DD_MM_YYYY.mat`.

2. In the local variables, instead of `mat` it will probably be easier to understand if you change it to `mat_10_10_2019`.

3. (needs further discussion @dqrobotics/developers ) It is strange that we are relying on two different mat files simultaneously, I strongly believe we need further discussion. My current understanding is that if we hope that the dual quaternion multiplication is within a threshold of 12 decimals (i.e. `DQ_threshold `), it does not seem particularly reasonable to expect that the multiplication among three dual quaternions will be within the same threshold.

Support information for 3.
- The [`assertEqual`](https://github.com/dqrobotics/python-tests/blob/4cd0f1ee9989c2ffb91a55bd743ecdfdb3751eaa/DQ_test.py#L87) used for testing multiplication does not have an explicit threshold but it uses the [DQ equality](https://github.com/dqrobotics/cpp/blob/93d4a05eb7ba1ab7c903f0710b76e2700d6a5fcc/src/DQ.cpp#L1302), which, itself, uses an elementwise `DQ_threshold.`

File: DQ_test.py [hunk 2]
```diff
@@ -24,8 +24,27 @@
...
+mat_feb_22_2023 = scipy.io.loadmat('DQ_test.mat')
```
Comment (mmmarinho) [inline DQ_test.py]: @juanjqo Hopefully the prior comment will also fix this, but otherwise I think that the filename and the variable name should be compatible.

File: DQ_test.py [hunk 3]
```diff
@@ -107,7 +126,7 @@ def test_ad(self):
     def test_adsharp(self):
         result_of_Adsharp = get_list_of_dq_from_mat('result_of_Adsharp', mat)
         for a, b, c in zip(dq_a_list, dq_b_list, result_of_Adsharp):
-            numpy.testing.assert_almost_equal(vec8(Adsharp(a, b)), vec8(c), 12, "Error in Ad")
+            numpy.testing.assert_almost_equal(vec8(Adsharp(a, b)), vec8(c), 12, "Error in Adsharp")
```
Comment (mmmarinho) [inline DQ_test.py]: @juanjqo  This might be a good time to replace all `12` by `DQ_threshold` instead.

**Review by mmmarinho — CHANGES_REQUESTED** [review]
@juanjqo A few changes and a discussion with @dqrobotics/developers .

File: DQ_test.py (hunk 1 above)
Comment (mmmarinho) [reply DQ_test.py]: @dqrobotics/developers For our information (FOI?):

https://www.mathworks.com/matlabcentral/answers/19736-identical-math-operations-in-c-and-matlab-differ-by-1-10-11-in-this-case

https://possiblywrong.wordpress.com/2017/09/12/floating-point-agreement-between-matlab-and-c/

File: DQ_test.py (hunk 2 above)
Comment (juanjqo) [reply DQ_test.py]: @mmmarinho I agree that in cases where there are several mat files, their names should be compatible. However, since we decided to use only one mat file for all tests, renaming it could be inconvenient and maybe unnecessary (?).
For instance, most of the tests in DQ_test.py read the mat file. Therefore, (please correct me if I'm wrong) each time we generate a new mat file, we must rename it in each test. 
Example:
Current:  (with mat_06_04_2023)
```python
# Binary operators
    def test_plus(self):
        result_of_plus = get_list_of_dq_from_mat('result_of_plus', mat_06_04_2023)
        for a, b, c in zip(dq_a_list, dq_b_list, result_of_plus):
            self.assertEqual(a + b, c, "Error in +")

    def test_minus(self):
        result_of_minus = get_list_of_dq_from_mat('result_of_minus', mat_06_04_2023)
        for a, b, c in zip(dq_a_list, dq_b_list, result_of_minus):
            self.assertEqual(a - b, c, "Error in -")

    def test_times(self):
        result_of_times = get_list_of_dq_from_mat('result_of_times', mat_06_04_2023)
        for a, b, c in zip(dq_a_list, dq_b_list, result_of_times):
            self.assertEqual(a * b, c, "Error in *")

    def test_dot(self):
        result_of_dot = get_list_of_dq_from_mat('result_of_dot', mat_06_04_2023)
        for a, b, c in zip(dq_a_list, dq_b_list, result_of_dot):
            self.assertEqual(dot(a, b), c, "Error in dot")
```
Let say that we generate a new mat file (mat_21_04_2023), in that case we would need to modify each test with the new name right?

```python
# Binary operators
    def test_plus(self):
        result_of_plus = get_list_of_dq_from_mat('result_of_plus', mat_21_04_2023)
        for a, b, c in zip(dq_a_list, dq_b_list, result_of_plus):
            self.assertEqual(a + b, c, "Error in +")

    def test_minus(self):
        result_of_minus = get_list_of_dq_from_mat('result_of_minus', mat_21_04_2023)
        for a, b, c in zip(dq_a_list, dq_b_list, result_of_minus):
            self.assertEqual(a - b, c, "Error in -")

    def test_times(self):
        result_of_times = get_list_of_dq_from_mat('result_of_times', mat_21_04_2023)
        for a, b, c in zip(dq_a_list, dq_b_list, result_of_times):
            self.assertEqual(a * b, c, "Error in *")

    def test_dot(self):
        result_of_dot = get_list_of_dq_from_mat('result_of_dot', mat_21_04_2023)
        for a, b, c in zip(dq_a_list, dq_b_list, result_of_dot):
            self.assertEqual(dot(a, b), c, "Error in dot")
```
I mean, it is not big deal, but it could be inconvenient. What do you think?

Best regards, 

Juancho

Comment (mmmarinho) [reply DQ_test.py]: @juanjqo I completely agree, in fact I think this is a perfect example of what I meant to say initially. 😊

**Comment by juanjqo** [comment]
@mmmarinho I implemented the required modifications. Please let me know if additional modifications are required. :neckbeard:

List of modifications: 

- There is only one mat file to all tests. Moreover, the mat file has a computer friendly name. 
- There are two thresholds related to DQ_threshold. The typical one (12) and the relaxed one (11). The latter is used in ad and adsharp tests only.

```python
# Threshold related to DQ_threshold
threshold = -(math.log(DQ_threshold)/math.log(10))

# Relaxed threshold used in some specific tests
relaxed_threshold = 0.1*threshold
```

Cheers, 

Juancho

**Comment by mmmarinho** [comment]
@juanjqo  How about

[`math.log10()`](https://docs.python.org/3/library/math.html#math.log10)

_Edit: also `relaxed_threshold = threshold - 1`_

File: DQ_test.py (hunk 1 above)
Comment (mmmarinho) [reply DQ_test.py]: @dqrobotics/developers I'll leave this last topic without `Resolve conversation` so that we can come back to it in due time.

For now, as @juanjqo mentioned, we'll use a relaxed threshold for `Ad` and `Adsharp`, so that we can move on with this and other improvements in the library.

We can come back to this with a more detail analysis and better solution in due course.

**Review by mmmarinho — APPROVED** [review]

**Comment by bvadorno** [comment]
Hi both,

From the thread, I couldn't follow the rationale behind different thresholds for `Ad` and `Adsharp`. Is this related to [this bug](https://github.com/dqrobotics/matlab/issues/68#issue-847315713)? Since I'm not following the exact implementation, I'm also puzzled by the definition of `threshold`. @juanjqo, may you clarify?

Bruno

**Comment by juanjqo** [comment]
Hi @bvadorno, I'm not sure if the bug you mentioned is related to this. 
The tests for Ad and Adsharp fails in Python/C++ but succeed in Matlab.
@mmmarinho [listed](https://github.com/dqrobotics/python-tests/pull/6#discussion_r1172086609) some possible explanations to the problem.

Context:
Recently, the operators Q4 and Q8 were added to [C++](https://github.com/dqrobotics/cpp/pull/56) and [Python](https://github.com/dqrobotics/python/pull/45). To include tests for both, I generated a new data set `DQ_test_06_04_2023.mat`. However, with the new data, the tests (python-tests/DQ_test.py) for `Ad` and `Adsharp` failed using a precision of 12  (i.e., `numpy.testing.assert_almost_equal(actual, desired, decimal=12)`).

As matter of fact, both tests for Ad and Ad Sharp succeed when a precision of 11 is used.

In the beginning, I proposed the use of both mat files (the old one, in which the tests for `Ad` and `Adsharp` were passing the tests with a precision of 12, and the new one `DQ_test_06_04_2023.mat`. The latter was only for the new tests (for Q4 and Q8).

After some iterations and suggestions proposed by @mmmarinho, the  PR was accepted with the following features:

- Only one mat file (the new one) is used for all tests. 
- The tests for `Ad` and `Adsharp` were relaxed. Instead of using a precision of `threshold=12`, we are using `relaxed_threshold=11`. (Only those tests are using `the relaxed_threshold`)
- Instead of using the numbers 12 and 11, both thresholds were defined as functions of DQ_threshold.
In other words, instead of
```python
threshold = 12
relaxed_threshold = 11
```
We have
```python
# Threshold related to DQ_threshold
threshold = -math.log10(DQ_threshold)

# Relaxed threshold used in some specific tests
relaxed_threshold = threshold-1
```
In this way, if `DQ_threshold` is modified, all thresholds used in `DQ_tests.py ` are updated automatically.

Best regards, 

Juancho
